### PR TITLE
Llama4Vision NNX

### PR DESCRIPTION
**Pull request by hengtaoguo**
# Description

Start with a short description of what the PR does and how this is a change from
the past.

The rest of the description includes relevant details and context, examples:
- why is this change being made,
- the problem being solved and any relevant context,
- why this is a good solution,
- some information about the specific implementation,
- shortcomings of the solution and possible future improvements.

If the change fixes a bug or a Github issue, please include a link, e.g.,:
FIXES: b/123456
FIXES: #123456

*Notice 1:* Once all tests pass, the "pull ready" label will automatically be assigned.
This label is used for administrative purposes. Please do not add it manually.

*Notice 2:* For external contributions, our settings currently require an approval from a MaxText maintainer to trigger CI tests.

# Tests

Please describe how you tested this change, and include any instructions and/or
commands to reproduce.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [ ] I have performed a self-review of my code.
- [ ] I have necessary comments in my code, particularly in hard-to-understand areas.
- [ ] I have run end-to-end tests tests and provided workload links above if applicable.
- [ ] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

File: src/MaxText/configs/models/llama4-17b-128e.yml [hunk 1]
```diff
@@ -41,4 +41,13 @@ temperature_tuning: True
...
+
+# Multimodal flags (need to set use_multimodal=true)
+image_size_for_vit: 336
+num_channels_for_vit: 3
+patch_size_for_vit: 14
+hidden_size_for_vit: 1408
+intermediate_size_for_vit: 5632
+num_hidden_layers_for_vit: 34
+num_attention_heads_for_vit: 16
+image_placeholder: <|image|>
```
Comment (bvandermoon) [inline src/MaxText/configs/models/llama4-17b-128e.yml]: Seems these are all the same values as in `base.yml`. Why not just read them from there instead? Same for the other file in this PR

File: src/MaxText/configs/models/llama4-17b-16e.yml [hunk 2]
```diff
@@ -41,4 +41,14 @@ temperature_tuning: True
...
+
+# Multimodal flags (need to set use_multimodal=true)
+image_size_for_vit: 336
+num_channels_for_vit: 3
+patch_size_for_vit: 14
+hidden_size_for_vit: 1408
+intermediate_size_for_vit: 5632
+num_hidden_layers_for_vit: 34
+num_attention_heads_for_vit: 16
+image_placeholder: <|image|>
```
Comment (bvandermoon) [inline src/MaxText/configs/models/llama4-17b-16e.yml]: Not sure if this matters but there are [quotes around this value in base.yml](https://github.com/AI-Hypercomputer/maxtext/blob/fafdeaa14183a8f5ca7b9f7b7542ce1655237574/src/MaxText/configs/base.yml#L808C19-L808C31):  `"<|image|>"`. Does it make a difference?

File: src/MaxText/layers/llama4.py [hunk 3]
```diff
@@ -130,104 +117,78 @@ def pixel_shuffle(input_tensor: Array, shuffle_ratio: float) -> Array:
...
+        in_features_shape=self.config.projector_input_dim_for_vit,
+        out_features_shape=self.config.projector_output_dim_for_vit,
+        dtype=self.config.dtype_mm,
         use_bias=False,
-        matmul_precision=cfg.matmul_precision,
+        matmul_precision=self.config.matmul_precision,
+        rngs=self.rngs,
     )
-    self.dropout = nn.Dropout(rate=cfg.projector_dropout_for_vit)
+    self.dropout = nnx.Dropout(rate=self.config.projector_dropout_for_vit, rngs=self.rngs)
```
Comment (bvandermoon) [inline src/MaxText/layers/llama4.py]: Can you use [`linears.Dropout`](https://github.com/AI-Hypercomputer/maxtext/blob/fafdeaa14183a8f5ca7b9f7b7542ce1655237574/src/MaxText/layers/linears.py#L292) instead here? There are known issues with `nnx.Dropout` in MaxText. Not sure why the unit tests were passing with this though, maybe they don't trigger this?

**Review by bvandermoon — COMMENTED** [review]
Thanks @hengtaoguo. Generally LGTM. Just a few comments/questions

File: src/MaxText/configs/models/llama4-17b-128e.yml (hunk 1 above)
Comment (hengtaoguo) [reply src/MaxText/configs/models/llama4-17b-128e.yml]: We added these values as default when we implemented Llama4. But afterwards, Gemma3 also uses these flags. To help better identify the difference in these two models (and for future models), I suggest we also attach those model-specific args here for readability. For example, [gemma3-4b.yml](https://github.com/AI-Hypercomputer/maxtext/blob/df8241562ed08d8e36ed39b5cc004e84f50ca704/src/MaxText/configs/models/gemma3-4b.yml#L35) also has these corresponding flags.

File: src/MaxText/configs/models/llama4-17b-16e.yml (hunk 2 above)
Comment (hengtaoguo) [reply src/MaxText/configs/models/llama4-17b-16e.yml]: Thanks! Updated the quotes here for consistency.

File: src/MaxText/layers/llama4.py (hunk 3 above)
Comment (hengtaoguo) [reply src/MaxText/layers/llama4.py]: Thanks for the catch! Updated to linears.Dropout and the outputs remain the same.

File: src/MaxText/layers/llama4.py [hunk 4]
```diff
@@ -751,42 +695,53 @@ def __call__(
...
     class_embedding_expanded = jnp.expand_dims(jnp.expand_dims(self.class_embedding, axis=0), axis=0)
-    class_embedding = jnp.broadcast_to(class_embedding_expanded, (hidden_states.shape[0], 1, cfg.hidden_size_for_vit))
+    class_embedding = jnp.broadcast_to(
+        class_embedding_expanded, (hidden_states.shape[0], 1, self.config.hidden_size_for_vit)
+    )
     hidden_states = jnp.concatenate([class_embedding, hidden_states], axis=1)
 
     # Add positional embedding
     hidden_states += self.positional_embedding_vlm
+    jax.debug.print("positional_embedding_vlm mean {}", jnp.mean(hidden_states))
```
Comment (aireenmei) [inline src/MaxText/layers/llama4.py]: Remove the debug.prints?

**Review by aireenmei — APPROVED** [review]
Thanks Hengtao?

File: src/MaxText/layers/llama4.py (hunk 4 above)
Comment (hengtaoguo) [reply src/MaxText/layers/llama4.py]: Thank you Aireen! All prints are removed.

**Comment by hengtaoguo** [comment]
> Newly generated checkpoint (with this NNX change) also works for decoding. [Command](https://paste.googleplex.com/4840988040364032), [workload](http://shortn/_6qQw16NNx1).

Tested again with the newly generated ckpt and works with a stable stack image: [command](https://paste.googleplex.com/4840988040364032), [workload](http://shortn/_6qQw16NNx1).

**Comment by hengtaoguo** [comment]
> Thanks! LGTM in general, wonder if you have tested the checkpoint conversion script?

Newly generated checkpoint (with this NNX change) also works for decoding. [Command](https://paste.googleplex.com/4840988040364032), [workload](http://shortn/_6qQw16NNx1).

**Comment by hengtaoguo** [comment]
> LGTM but can you please collect some before/after profiles so we can ensure that the HLOs are the same before/after this change?

Thanks! I generated the profiles with/without this change and the memory profile seems not impacted (using the same [command](https://paste.googleplex.com/6628439243227136)):

- Before this change: http://shortn/_b19Dv8RJ4B, [logs](https://pantheon.corp.google.com/logs/query;query=resource.type%3D%22k8s_container%22%0Aresource.labels.project_id%3D%22cloud-tpu-multipod-dev%22%0Aresource.labels.location%3D%22europe-west4%22%0Aresource.labels.cluster_name%3D%22v5p-128-bodaborg-europe-west4-b%22%0Aresource.labels.namespace_name%3D%22default%22%0Aresource.labels.pod_name:%22hengtaoguo-2025-10-06-20-48-28-slice-job-0-0-%22%20severity%3E%3DDEFAULT;storageScope=project;duration=P1D?e=13802955&mods=allow_workbench_image_override&project=cloud-tpu-multipod-dev)
- After this change: http://shortn/_K5oTH30V9S, [logs](https://pantheon.corp.google.com/logs/query;query=resource.type%3D%22k8s_container%22%0Aresource.labels.project_id%3D%22cloud-tpu-multipod-dev%22%0Aresource.labels.location%3D%22europe-west4%22%0Aresource.labels.cluster_name%3D%22v5p-128-bodaborg-europe-west4-b%22%0Aresource.labels.namespace_name%3D%22default%22%0Aresource.labels.pod_name:%22hengtaoguo-2025-10-06-20-46-24-slice-job-0-0-%22%20severity%3E%3DDEFAULT;storageScope=project;duration=P1D?e=13802955&mods=allow_workbench_image_override&project=cloud-tpu-multipod-dev)
